rfcs/002 - Database Access.md

```diff
@@ -0,0 +1,117 @@
+
+## Proposal
+
+Just alternatives right now as we dig into the possibilities here. The solution ideally should:
+
+1. Declare what database will be used by the application, whether a local sqlite file, a remote Postgres/MySQL/etc server, or perhaps something more esoteric like MongoDB, Redis, etc.
+2. Declare the connection is read-only vs read-write, so it *never* tries to mutate a read-only database connection.
+3. Allow the user to define the schema within Marsha if the database is *owned* by the Marsha application (one level above read-write, as it should also automatically migrate the database on startup ideally)
+4. When run as a web server, it should use a singleton connection instead of creating a new connection per request, since many databases handle large connection volumes poorly. (Likely the CLI approach could use the same mechanism, though it wouldn't matter there)
+5. Multiple databases should be allowed in the same marsha application (the syntax shouldn't be a singleton type) so there needs to be a clear way to associate schemas and queries to a particular database
+6. When possible, get the answer from the database during compilation, instead of making the user explicitly spell it out in Marsha text.
+
+### Alternatives Considered
+
+#### Fully separate syntax
+
+Making a wholly-separate syntax added into Marsha: `database`, `query`, and `schema` blocks. The example syntaxes below are vaguely based on the syntax explanations [Postgres uses](https://www.postgresql.org/docs/current/sql-select.html). `{something}` is a value the user needs to provide and `[ something ]` is an optional part of the syntax. Unbracketed/unbraced words are keywords. But outside of the more rigid syntactic areas the example data is just that -- example data.
+
+```md
+# database {name} {url or path} [ro | rw]
+```
+
+Declares the database to be used, names it, and specifies the URL or path to the sqlite file. Optionally specifies read-only or read-write, defaulting to read-only.
+
+```md
+# query {name}(args) using {dbname}
+
+Query description
+
+---
+return column 1, return column 2, return column 3
+example 1 value 1, example 1 value 2, example 1 value 3
+example 2 value 1, example 2 value 2, example 2 value 3
+example 3 value 1, example 3 value 2, example 3, value 3
+```
+
+Similar to a function, but specifying which database it is connected to with `using [dbname]` and the return type is a custom type defined at the end after a horizontal rule `---`.
```

Comment:
I think theres a typo here since the dbname would not be optional?
```suggestion
Similar to a function, but specifying which database it is connected to with `using {dbname}` and the return type is a custom type defined at the end after a horizontal rule `---`.
```